Refactor the API design for `cmor_handlers`

tomvothecoder opened the pull request:
Summary of Changes
- Closes #103 

Related issues
- https://github.com/E3SM-Project/e3sm_to_cmip/issues/139
- https://github.com/E3SM-Project/e3sm_to_cmip/issues/129


## Solution Implementation

### Phase 1: Design and implement classes
- [x] Identify common design patterns that are redundant in the existing codebase
- [x] Create skeleton classes with attributes and methods based on common design patterns in the existing code
  - [x] BaseVarHandler, VarHandler
- [x] Define class methods and attributes using existing implementations
- [x] Write set of base tests

### Phase 2: Refactor how handlers are defined (non-MPAS vars)
- [x] Organize handlers in sub-directories (MPAS and regular handlers)
- [x] Define all handlers in a single `handlers.yaml` file -- use Python modules and `default_handlers_info.yaml`
- [x] Define handler formula functions in a `formulas.py` file ([requirement 1](https://docs.google.com/document/d/17mcv0OLXAg3smOVosuCCKf8a-4H8zuOwfX3I2REohfQ/edit#bookmark=id.83w6mic8a7fy))
- [x] Handle when there are multiple formulas for a handler -- simply check the dictionary keys that stores the arrays contain the required key(s) for the formula
- [x] Determine if all handlers should nans with FILL_VALUE (only some formula functions do at the moment)

### Phase 3: Refactor how handlers are derived

Note, these methods remove the `--var-list "all"` capability, which is an implicit way of attempting to derive CMIP handlers based on the available E3SM input variables. **This is a breaking change for those who use this feature.** We probably expected `--var-list "all"` isn't used because CWL workflows are available and they explicitly pass `--varlist` options.

- [x] Add utility function for loading handlers (info mode)
- [x] Add utility function for deriving handlers based on input E3SM datasets ([requirement 2](https://docs.google.com/document/d/17mcv0OLXAg3smOVosuCCKf8a-4H8zuOwfX3I2REohfQ/edit#bookmark=id.cwvxkq1gspgq))

```
# Pseudocode for deriving handlers.
1. Get a dictionary of all available handler objects. 
  - Key = CMIP variable id,  value = list of available handler objects defined in `handlers.yaml`
	
  - Example: 
	available_handlers = {
            "pr": [
                VarHandler(name="pr", raw_variables=["PRECT"]),
                VarHandler(name="pr", raw_variables=["PRECC", "PRECL"]),
            ],
        }
2. Loop over user specified CMIP variable ids (e2c `--var-list` option)
  - Get the list of handlers from `available_handlers` dict
  - Derive the appropriate handler based on the available E3SM variable id(s) from the input file(s)
  - Raise error if matching handler can’t be derived
3. Return list of matching handlers
```
- [x] Keep and load handlers for `areacella`, `clisccp`, `orog`, `pfull`, `phalf`, and all MPAS vars.
  - These 5 specific handlers need to be refactored because they contain legacy code and/or rely on `cdms2` and `cdutil`
  - MPAS vars use different processing methods and would need to be revisted if we wanted to also refactor them.		     
### Phase 4: Refactor how `e3sm_to_cmip` is executed internally  (convert `__main__.py` into a Python class)
This allows for easier debugging and testing, since we can use the interactive console, debugging, etc. (provide an example script for developers).
We won't need to run `e3sm_to_cmip` on the command line every time there is a code change.
- [x] Add class `E3SMtoCMIP`
- [x] Move methods for parsing CLI arguments into `E3SMtoCMIP`
- [x] Parse CLI arguments into `E3SMtoCMIP` attributes
- [x] Run `load_handlers()` for `info_mode` and `derive_handlers()` for all other modes

### Phase 5: Code validation and regression testing
- [x] Add unit tests for new classes and functions
  - Some methods in VarHandler are skipped for now because they wrap existing functions that need to be refactored
- [x] Setup common directory for e3sm_to_cmip_testing (`/lcrc/group/e3sm/e3sm_to_cmip`)
  - [x] git clone https://github.com/PCMDI/cmip6-cmor-tables
  - [x] git clone https://github.com/E3SM-Project/CMIP6-Metadata
  - [x] How the underlying shared testing library is structured:
	```
	/lcrc/group/e3sm/e3sm_to_cmip
	  - /cmip6-cmor-tables
	  - /CMIP6-Metadata
	  - /test-cases
	  - # dataset from each realm
	    - /atm-unified-eam
	      - /input-regridded  -- stores regridded input data for CMORizing
	      - /input-raw --  <OPTIONAL> stores raw input data before regridding
	      - /output-103-refactor-cmor-handlers -- stores CMORized output from the branch
	      - /output-master  -- stores CMORized output from master
	    - /lnd-elm
	       ...
	    - /mpaso
	       ...
	    - mpassi
	      ...
	```
- [x] Add regridded (with nco) input files to each test case's `/input-regridded` (via @chengzhuzhang)
  -Related to #174
- [x] Regression testing script
	- [x] Perform complete test runs using `master` branch 
	  -  Store outputs in `/lcrc/group/e3sm/e3sm_to_cmip/test-cases/master`
	- [x] Perform complete test runs using `103-refactor-cmor-handlers` branch
	  -  Store outputs in `/lcrc/group/e3sm/e3sm_to_cmip/test-cases/103-branch`
- [x] Dataset validation -- related scripts and results stored in `/scripts/103-branch-regression-test`
	- [x] Loop over output datasets and check identical with xarray

### Phase 6: Documentation and Cleanup
- [x] Update documentation with how to add new handlers
  - [x] Convert README.md to README.rst and remove redundant installation section
  - [x] Add getting started page with installation instructions
  - [x] Add a page describing variable handlers. What are variable handlers? How do I add them?
  - [x] Update usage sections with latest --help text (grouped args)
  - [x] Add development guide
- [x] Delete regular handler Python modules and `default_handlers_info.yaml` (do this at the end of this PR)
- [x] Delete unused functions such as `_load_handlers()`
- [x] Improve logging 
  - [x] Improve format -- try to get close to what it was before
  - [x] Add logging for helpful outputs
  - [x] Add line spaces in between CMORizing runs

## Future Work
* Refactor `VarHandler` methods and write tests
* Refactor organization of functions into `VarHandler` and `E3SMtoCMIP` classes
* Enforce checks on CLI arguments (avoid silent errors)
* Remove unused arguments

tomvothecoder commented:
Hey @chengzhuzhang, should `np.nan` be replaced with `FILL_VALUE` for **all handlers**?

I noticed from legacy code that some handlers do this, while others don't.

Example: https://github.com/E3SM-Project/e3sm_to_cmip/pull/140/files#diff-fd01a882c31af6aeedf6a554b47e4b0f18db0630f992ca8f976036702304f600R16-R26

chengzhuzhang commented:
> Hey @chengzhuzhang, should for **all handlers** replace np.nan`with`FILL_VALUE`?
> 
> I noticed from legacy code that some handlers do this, while others don't.
> 
> Example: https://github.com/E3SM-Project/e3sm_to_cmip/pull/140/files#diff-fd01a882c31af6aeedf6a554b47e4b0f18db0630f992ca8f976036702304f600R16-R26

Thanks for catching this. I believe we should replace np.nan with FILL_VALUE for all handlers.

chengzhuzhang commented:
@tomvothecoder I'm creating datasets for code validation and regression testing and placing data under /lcrc/group/e3sm/e3sm_to_cmip/test-cases/, but running into permission issue in the target directory. Could you help me to fix the permission? Or I can go ahead and create the folders on acme1 and you could help to copy over?

tomvothecoder commented:
> @tomvothecoder I'm creating datasets for code validation and regression testing and placing data under /lcrc/group/e3sm/e3sm_to_cmip/test-cases/, but running into permission issue in the target directory. Could you help me to fix the permission? Or I can go ahead and create the folders on acme1 and you could help to copy over?

Thanks for creating the datasets Jill. I just updated the permissions for ` /lcrc/group/e3sm/e3sm_to_cmip/test-cases/` to include `E3SM` group ownership with `rwx` privileges. Let me know if this fixes the permission issue.

```bash
(base) [ac.tvo@chrlogin1]/lcrc/group/e3sm/e3sm_to_cmip% ls -al
total 36
drwxrws---+   5 ac.tvo E3SM  4096 Oct 17 13:22 .
drwxrwsr-x+ 189 root   E3SM 16384 Nov 27 17:55 ..
drwxr-s---+   7 ac.tvo E3SM  4096 Oct  4 15:59 cmip6-cmor-tables
drwxr-s---+   7 ac.tvo E3SM  4096 Oct  4 16:00 CMIP6-Metadata
drwxrwxr-x    6 ac.tvo E3SM  4096 Oct 17 17:53 test-cases
```

tomvothecoder commented:
Hey @chengzhuzhang, I uploaded the validation scripts and the results here: https://github.com/E3SM-Project/e3sm_to_cmip/tree/feature/103-refactor-cmor-handlers/scripts/103-branch-regression-test.

Validation [results](https://github.com/E3SM-Project/e3sm_to_cmip/blob/feature/103-refactor-cmor-handlers/scripts/103-branch-regression-test/2023-02-06T16-47-01-output-comparison.csv):
* All dataset files are identical (minus the `creation_date`, `history`, `tracking_id` as expected).
* The `clw` has very small floating differences that are not of concern. I wrote a separate [script](https://github.com/E3SM-Project/e3sm_to_cmip/blob/feature/103-refactor-cmor-handlers/scripts/103-branch-regression-test/debug-clw-debug-diffs.py) to debug this.

tomvothecoder commented on scripts/example_end-to-end_script.sh:
```diff
@@ -1,8 +1,15 @@
 #!/bin/bash
```
I formatted `example_end-to-end_script.sh` to make it a bit easier to read.

tomvothecoder commented on setup.py:
```diff
@@ -38,13 +39,13 @@ def run(self):
-                "using the Climate Model Output Rewriter.",
-    entry_points={'console_scripts':
-                  ['e3sm_to_cmip = e3sm_to_cmip.__main__:main']},
-    packages=['e3sm_to_cmip', 'e3sm_to_cmip.cmor_handlers'],
```
I updated the `packages` value in `setup` because it was breaking. The standard convention based on what I found is to use `find_packages()` instead of statically defining the included modules.

tomvothecoder reviewed (commented):
My self-review of this massive PR.

chengzhuzhang commented:
Awesome updates! Thanks for getting the validation scrips in. It turned out the following link is not available though. 
> * [script](https://github.com/E3SM-Project/e3sm_to_cmip/blob/feature/103-refactor-cmor-handlers/scripts/103-branch-regression-test/debug-clw-debug-diffs.py)
I'm very much looking forward to have this PR merged!

tomvothecoder commented:
> Awesome updates! Thanks for getting the validation scrips in. It turned out the following link is not available though.

I fixed the link in original [comment above](https://github.com/E3SM-Project/e3sm_to_cmip/pull/140#issuecomment-1420017549).

I recommend you and @TonyB9000 check out these documentation pages to make sure they are clear.
I'll update or fix things if needed.
- https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html
- https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/development-guide.html

chengzhuzhang commented on docs/source/development-guide.rst:
```diff
@@ -0,0 +1,106 @@
+
+In older versions of ``e3sm_to_cmip``, the only way to test changes was to run ``e3sm_to_cmip`` directly on the command line. This debugging process often involves adding ``print`` or ``ipdb`` statements throughout the codebase, which generally isn't good practice because it is inefficient and developers might forget to delete those statements.
+
+As of ``e3sm_to_cmip > 1.91``,  ``e3sm_to_cmip`` can now be executed through a Python script Advantages of using this approach include:
```
Missing a  period at the end of this line.

chengzhuzhang commented on docs/source/development-guide.rst:
```diff
@@ -0,0 +1,106 @@
+
+.. code-block:: bash
+
+    e3sm_to_cmip --info -v all --input /p/user_pub/work/E3SM/1_0/historical/1deg_atm_60-30km_ocean/atmos/native/model-output/day/ens1/v1/ --tables /home/vo13/PCMDI/cmip6-cmor-tables/Tables/
```
Should it be -v 'prw, pr'? not sure if `all` is still supported..

tomvothecoder replied:
Good catch and yes `all` is removed.

chengzhuzhang commented on docs/source/development-guide.rst:
```diff
@@ -0,0 +1,106 @@
+    args = [
+        "--info",
+        "-v",
+        "prw pr",
```
Not sure if a comma is needed here, i.e., "prw, pr"

tomvothecoder replied:
I tested with and without a comma and they both work.

I updated with comma because it looks cleaner.

chengzhuzhang commented on docs/source/getting-started.rst:
```diff
@@ -0,0 +1,122 @@
+
+      git clone https://github.com/E3SM-Project/e3sm_to_cmip.git
+      cd e3sm_to_cmip
+      conda env create -f conda/dev.yml
```
`conda env create -f conda-env/dev.yml` this has been updated on master.

tomvothecoder replied:
Fixed

chengzhuzhang commented on docs/source/var-handlers.rst:
```diff
@@ -0,0 +1,155 @@
+            levels: null
+
+2. ``e3sm_to_cmip/cmor_handlers/vars`` directory
+    - Each handler is defined as Python module (``.py`` file). This is the legacy design for  defining handlers, which is progressively being be replaced by ``handlers.yaml``.
```
Maybe change to `which is progressively being replaced by`

tomvothecoder replied:
Fixed

chengzhuzhang commented on docs/source/var-handlers.rst:
```diff
@@ -0,0 +1,155 @@
+
+For example, let's say we want to CMORize the variable ``"pr"`` and we pass an E3SM input dataset that has the variables ``"PRECC"`` and ``"PRECL"``. ``e3sm_to_cmip`` derives the appropriate ``"pr"`` variable handler using this logic flow:
+
+1. Run ``e3sm_to_cmip --var-list pr --input-path <SOME_INPUT_PATH>``
```
I'm a little bit confused about type of --var-list as input, in the helper script it says`Space separated list of variables`, but I saw instances of using, pr, 'pr', ['pr'], I suppose all three are supported as input?

tomvothecoder replied:
``--varlist`` accepts a list of strings with or without a comma (`pr, prw` or `pr prw` ). I prefer with a comma though because it looks a cleaner.

chengzhuzhang commented:
> I recommend you and @TonyB9000 check out these documentation pages to make sure they are clear. I'll update or fix things if needed.
> 
>     * https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html
> 
>     * https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/development-guide.html

Hey @tomvothecoder the docs are nice and clean! Easy for me to understand. The logic for the new variable definition and derivation is clean and is very well explained! I only have very minor comments/questions.

tomvothecoder commented:
> > I recommend you and @TonyB9000 check out these documentation pages to make sure they are clear. I'll update or fix things if needed.
> > ```
> > * https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html
> > 
> > * https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/development-guide.html
> > ```
> 
> Hey @tomvothecoder the docs are nice and clear! Easy for me to understand. The logic for the new variable definition and derivation is clear and is very well explained! I only have very minor comments/questions.

Thanks for the review Jill! I addressed your comments.

I also split up sections for working with atmosphere/land variables and MPAS variables. MPAS variables are derived differently so it made sense to split them up for clarity.

* https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html#working-with-atmosphere-and-land-variables
* https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html#working-with-mpas-ocean-and-sea-ice-variables
  * https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html#how-e3sm-to-cmip-derives-mpas-variable-handlers

**After I clean up the Python logging and your final review, it will be ready to merge.**

TonyB9000 commented:
This is good practice, as an additional entry made (at the end of a list) need not worry that a necessary comma added to the end of the previous last entry be added.

Good that it works both ways. (C compilers should allow the same – although I don’t know that they do.)


From: Tom Vo ***@***.***>
Sent: Thursday, February 9, 2023 10:14 AM
To: E3SM-Project/e3sm_to_cmip ***@***.***>
Cc: Bartoletti, Tony ***@***.***>; Mention ***@***.***>
Subject: Re: [E3SM-Project/e3sm_to_cmip] Refactor the API design for `cmor_handlers` (PR #140)


@tomvothecoder commented on this pull request.

________________________________

In docs/source/development-guide.rst<https://urldefense.us/v3/__https:/github.com/E3SM-Project/e3sm_to_cmip/pull/140*discussion_r1101854861__;Iw!!G2kpM7uM-TzIFchu!myjZR70UGWnyuTZPMBfkI5rlu_Oga1daREQi5hxOMmbljb5hiKq_zzMQZdqIlQ$>:

> +CLI Execution

+

+.. code-block:: bash

+

+    e3sm_to_cmip --info -v all --input /p/user_pub/work/E3SM/1_0/historical/1deg_atm_60-30km_ocean/atmos/native/model-output/day/ens1/v1/ --tables /home/vo13/PCMDI/cmip6-cmor-tables/Tables/

+

+Python Execution

+

+.. code-block:: python

+

+    from e3sm_to_cmip.__main__ import main

+

+    args = [

+        "--info",

+        "-v",

+        "prw pr",

I tested with and without a comma and they both work.

I updated with comma because it looks cleaner.

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https:/github.com/E3SM-Project/e3sm_to_cmip/pull/140*discussion_r1101854861__;Iw!!G2kpM7uM-TzIFchu!myjZR70UGWnyuTZPMBfkI5rlu_Oga1daREQi5hxOMmbljb5hiKq_zzMQZdqIlQ$>, or unsubscribe<https://urldefense.us/v3/__https:/github.com/notifications/unsubscribe-auth/ANZBCGQTU4QSVARM2JOXWETWWUXXRANCNFSM5YVXAOPQ__;!!G2kpM7uM-TzIFchu!myjZR70UGWnyuTZPMBfkI5rlu_Oga1daREQi5hxOMmbljb5hiKq_zzNAnLItfA$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

TonyB9000 commented:
@Vo, ***@***.***>:

            After I clean up the Python logging and your final review, it will be ready to merge

I once added some logging that reiterates input values look for “e3sm_to_cmip.util-<timestamp>” print-outs.  Those could be eliminated, or only triggered via a “—verbose” flag, etc.



From: Tom Vo ***@***.***>
Sent: Thursday, February 9, 2023 10:19 AM
To: E3SM-Project/e3sm_to_cmip ***@***.***>
Cc: Bartoletti, Tony ***@***.***>; Mention ***@***.***>
Subject: Re: [E3SM-Project/e3sm_to_cmip] Refactor the API design for `cmor_handlers` (PR #140)


I recommend you and @TonyB9000<https://urldefense.us/v3/__https:/github.com/TonyB9000__;!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm5rxMC9xg$> check out these documentation pages to make sure they are clear. I'll update or fix things if needed.

* https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html<https://urldefense.us/v3/__https:/e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html__;!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm6bsJzIrQ$>



* https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/development-guide.html<https://urldefense.us/v3/__https:/e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/development-guide.html__;!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm4myMzizA$>

Hey @tomvothecoder<https://urldefense.us/v3/__https:/github.com/tomvothecoder__;!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm5MVymBvQ$> the docs are nice and clear! Easy for me to understand. The logic for the new variable definition and derivation is clear and is very well explained! I only have very minor comments/questions.

Thanks for the review Jill! I addressed your comments.

I also split up sections for working with atmosphere/land variables and MPAS variables. MPAS variables are derived differently so it made sense to split them up for clarity.

  *   https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html#working-with-atmosphere-and-land-variables<https://urldefense.us/v3/__https:/e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html*working-with-atmosphere-and-land-variables__;Iw!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm7hfO5eUQ$>
  *   https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html#working-with-mpas-ocean-and-sea-ice-variables<https://urldefense.us/v3/__https:/e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html*working-with-mpas-ocean-and-sea-ice-variables__;Iw!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm40zHtmkQ$>
     *   https://e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html#how-e3sm-to-cmip-derives-mpas-variable-handlers<https://urldefense.us/v3/__https:/e3sm-to-cmip.readthedocs.io/en/feature-103-refactor-cmor-handlers/var-handlers.html*how-e3sm-to-cmip-derives-mpas-variable-handlers__;Iw!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm4ETMfMVw$>

After I clean up the Python logging and your final review, it will be ready to merge.

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https:/github.com/E3SM-Project/e3sm_to_cmip/pull/140*issuecomment-1424625563__;Iw!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm5EUZL9ZQ$>, or unsubscribe<https://urldefense.us/v3/__https:/github.com/notifications/unsubscribe-auth/ANZBCGVL2LEFOQHQS77AQTDWWUYKJANCNFSM5YVXAOPQ__;!!G2kpM7uM-TzIFchu!ngy2EOOxe8eThO4l0iVb76p2IRxI7lZEO_lujlYBNglRJyb7U7y4Qm6PGTdo0g$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

tomvothecoder commented:
I think this is ready to be merged now. 

All of the print statements have been replaced with the Python logging module, which logs messages to a logfile and console. It works for both serial and parallel runs.

I added helpful information such as the configuration, input dataset variables, derived variables, etc. I also removed styled messages because text files don't support color.

Example:

```bash
(e2c_dev) bash-4.2$ e3sm_to_cmip -i /p/user_pub/e3sm/e3sm_to_cmip/test-cases/103-branch/diff/rgr -o /p/user_pub/e3sm/e3sm_to_cmip/test-cases/103-branch/diff -v rlus, clw -t /p/user_pub/e3sm/zhang40/e3sm_to_cmip_data/cmor/cmip6-cmor-tables/Tables/ -u /p/user_pub/e3sm/zhang40/e3sm_to_cmip_data/user_metadata.json --serial
2023-02-09 19:53:18,957_957:INFO:__init__:--------------------------------------
2023-02-09 19:53:18,958_958:INFO:__init__:| E3SM to CMIP Configuration
2023-02-09 19:53:18,958_958:INFO:__init__:--------------------------------------
2023-02-09 19:53:18,959_959:INFO:__init__:    * var_list='['rlus', 'clw']'
2023-02-09 19:53:18,959_959:INFO:__init__:    * input_path='/p/user_pub/e3sm/e3sm_to_cmip/test-cases/103-branch/diff/rgr'
2023-02-09 19:53:18,959_959:INFO:__init__:    * output_path='/p/user_pub/e3sm/e3sm_to_cmip/test-cases/103-branch/diff'
2023-02-09 19:53:18,960_960:INFO:__init__:    * precheck_path='None'
2023-02-09 19:53:18,960_960:INFO:__init__:    * freq='mon'
2023-02-09 19:53:18,961_961:INFO:__init__:    * realm='atm'
2023-02-09 19:53:20,507_507:INFO:_get_handlers:--------------------------------------
2023-02-09 19:53:20,508_508:INFO:_get_handlers:| Derived CMIP6 Variable Handlers
2023-02-09 19:53:20,508_508:INFO:_get_handlers:--------------------------------------
2023-02-09 19:53:20,508_508:INFO:_get_handlers:    * 'rlus' -> ['FLDS', 'FLNS']
2023-02-09 19:53:20,508_508:INFO:_get_handlers:    * 'clw' -> ['CLDLIQ']
2023-02-09 19:53:20,508_508:INFO:run:Writing log output to: logs/20230209_195318_951625
2023-02-09 19:53:20,509_509:INFO:run:--------------------------------------
2023-02-09 19:53:20,510_510:INFO:run:| Running E3SM to CMIP in Serial
2023-02-09 19:53:20,510_510:INFO:run:--------------------------------------
2023-02-09 19:53:20,510_510:INFO:run_serial:Trying to CMORize with handler: {'name': 'rlus', 'units': 'W m-2', 'raw_variables': ['FLDS', 'FLNS'], 'table': 'CMIP6_Amon.json', 'unit_conversion': None, 'formula': 'FLDS + FLNS', 'formula_method': <function rlus at 0x7feb7de37910>, 'positive': 'up', 'levels': None, 'output_data': None, 'method': <bound method VarHandler.cmorize of <e3sm_to_cmip.cmor_handlers.handler.VarHandler object at 0x7feb7dcf7760>>}
2023-02-09 19:53:20,511_511:INFO:handle_variables:rlus: Starting
2023-02-09 19:53:20,730_730:INFO:handle_variables:rlus: CMOR setup complete
2023-02-09 19:53:20,730_730:INFO:handle_variables:rlus: loading FLDS
2023-02-09 19:53:20,741_741:INFO:handle_variables:rlus: loading FLNS
2023-02-09 19:53:20,751_751:INFO:handle_variables:rlus: loading axes
2023-02-09 19:53:20,754_754:INFO:handle_variables:rlus: time 0.0 - 365.0
rlus: time 0.0 - 365.0: 100%|████████████████████████████████| 12/12 [00:00<00:00, 98.60it/s]
2023-02-09 19:53:21,168_168:INFO:run_serial:Finished rlus, 1/2 jobs complete
2023-02-09 19:53:21,170_170:INFO:run_serial:Trying to CMORize with handler: {'name': 'clw', 'units': 'kg kg-1', 'raw_variables': ['CLDLIQ'], 'table': 'CMIP6_Amon.json', 'unit_conversion': None, 'formula': None, 'positive': None, 'levels': {'name': 'standard_hybrid_sigma', 'units': '1', 'e3sm_axis_name': 'lev', 'e3sm_axis_bnds': 'ilev'}, 'output_data': None, 'method': <bound method VarHandler.cmorize of <e3sm_to_cmip.cmor_handlers.handler.VarHandler object at 0x7feb7dcf7130>>}
2023-02-09 19:53:21,170_170:INFO:handle_variables:clw: Starting
2023-02-09 19:53:21,274_274:INFO:handle_variables:clw: CMOR setup complete
2023-02-09 19:53:21,274_274:INFO:handle_variables:clw: loading CLDLIQ
2023-02-09 19:53:22,845_845:INFO:handle_variables:clw: loading axes
2023-02-09 19:53:22,852_852:INFO:handle_variables:clw: time 0.0 - 365.0
clw: time 0.0 - 365.0: 100%|█████████████████████████████████| 12/12 [00:08<00:00,  1.43it/s]
2023-02-09 19:53:32,055_055:INFO:run_serial:Finished clw, 2/2 jobs complete
2023-02-09 19:53:32,055_055:INFO:run_serial:2 of 2 handlers complete

```

TonyB9000 commented:
Excellent on the logging!

chengzhuzhang reviewed (approved):
Looking great. I think we can forge ahead with the merge!
